### PR TITLE
Feature: limited datatypes

### LINKED ISSUE #336: Some operations on `xsd:Integer` are unecessarily expensive
For example: casting an `xsd:Integer` to string requires the following in most cases
1. un-inline the integer value into a int64_t (this is true for most integers)
2. create a `boost::multiprecision` integer out of it
3. call the to string function for the `boost::multiprecision` integer

In theory it would be more optimal if we did not construct the boost integer, instead just converting the uninlined int64_t to string.

Not sure on the implementation yet.

### LINKED ISSUE #369: add size limits for unlimited datatypes
currently, the size of an integer is only limited by the available memory, which could lead to problems.
add a fixed size limit, generously over the limits required by the specifications.

- [x] integer (including negative, positive and non-negative, non-positive)
- [x] decimal
- [ ] also document this properly
- [ ] conversion double→integer

XSD required datatype ranges: https://www.w3.org/TR/xmlschema11-2/#partial-implementation

- for Decimal with $i/10^k$, 64-bit integer is enough for $i$ and for $k$ 1 Byte would be enough. ($i < 10^{16} < 2^{63}$)
- Conversion from double to decimal needs at least 1024 bit integer to be lossless.
    - boost fixed size 128 bit integer does not throw exceptions on overflow (most likely fallback to [[__int128](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fint128.html)](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fint128.html))

RDF4CPP should implement:

- arbitrary precision integer types: `__int128` and `unsigned __int128`
- decimals: i: 128 bit, k: 64 bit
- for conversion double→integer: temp 1024 bit integer, throw if 128 bit integer is exceeded

### PR DESCRIPTION
**Pull request by mcb5637**
close #369 

TODO:
- [x] specialized node storages
- [x] fallback to boost if __int128 is not available
- [x] #336 (__int128 is faster to construct)
- [x] documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -113,11 +113,34 @@ cd build_dir
...
+
+For `http://www.w3.org/2001/XMLSchema#duration` (and all derived types), both parts have a separate signed 64-bit integer
+and with it its associated limits. The seconds part of the duration supports nanosecond precision.
+
 ### Additional CMake config options:
 
 - `-DBUILD_EXAMPLES=ON/OFF [default: OFF]`: Build the examples.
 - `-DBUILD_TESTING=ON/OFF [default: OFF]`: Build the tests.
 - `-DBUILD_SHARED_LIBS=ON/OFF [default: OFF]`: Build a shared library instead of a static one.
+- `-DRDF4CPP_USE_UNLIMITED_DATATYPES=ON`: If defined, uses unlimited datatypes for integer and decimal types.
```
Comment (liss-h) [inline README.md]: I don't think we need this. Its more maintenance overhead for us and we won't use it

**Review by liss-h — CHANGES_REQUESTED** [review]
First mini review: I didn't really look at anything, but the one thing I noticed is that you added an option to stay unlimited.

For the benefit of maintenance we should only support the limited integers.

Sorry that we weren't clear enough on that

File: CMakeLists.txt [hunk 2]
```diff
@@ -178,6 +178,17 @@ target_link_libraries(rdf4cpp
         highway::highway
         )
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
```
Comment (liss-h) [inline CMakeLists.txt]: Since this is only needed on Linux ARM (if possible) please also make this conditional on the target

File: src/rdf4cpp/util/Int128.hpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+#include <rdf4cpp/writer/BufWriter.hpp>
+#include <dice/hash.hpp>
+
+namespace rdf4cpp::util {
+#ifdef __SIZEOF_INT128__
+    using Int128 = __int128;
+#else
+    using Int128 = boost::multiprecision::checked_int128_t;
+#endif
+    using BigInt = Int128;
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: Unecessary typdef

File: src/rdf4cpp/util/Int128.hpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        using cpp_int_unchecked = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<MinBits, MaxBits, SignType, boost::multiprecision::unchecked, Alloc>>;
+
+        template<typename T>
+        concept IntegralExt = std::integral<T> || std::same_as<T, __int128> || std::same_as<T, unsigned __int128>;
+        template<typename T>
+        concept BoostNumber = std::same_as<T, boost::multiprecision::number<typename T::backend_type, T::et>>;
+
+        template<OverflowMode m, typename T>
+        requires IntegralExt<T>
+        static constexpr bool add_checked(T const &a, T const &b, T &result) noexcept {
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: I think it would make sense to add a `CheckedInt` specialization for boost mp types instead of putting all the functions here

File: src/rdf4cpp/util/Int128.hpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    using Int128 = __int128;
+#else
+    using Int128 = boost::multiprecision::checked_int128_t;
+#endif
+    using BigInt = Int128;
+
+    // string -> int128 can be found in CharConvExt
+
+#ifdef __SIZEOF_INT128__
+    inline bool to_chars_canonical(__int128 value, writer::BufWriterParts const writer) noexcept {
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: I have a suggestion for this. Since I would assume the standard library should be pretty good at serializing integers we could utilize `std::to_chars` for this and avoid a bunch of math (maybe). Instead of dividing by 10 we divide by the largest power of 10 that fits into a 64 bit integer:

```c++
inline constexpr unsigned __int128 max_pow10 = 10'000'000'000'000'000'000ull;

std::to_chars_result to_chars(char *first, char *last, unsigned __int128 val) noexcept {
    if (val <= std::numeric_limits<uint64_t>::max()) {
        return std::to_chars(first, last, static_cast<uint64_t>(val));
    }

    std::to_chars_result upper = to_chars(first, last, val / max_pow10);
    if (upper.ec != std::errc{}) {
        return upper;
    }

    first = upper.ptr;

    std::to_chars_result lower = std::to_chars(first, last, static_cast<uint64_t>(val % max_pow10));
    return lower;
}

std::to_chars_result to_chars(char *first, char *last, __int128 val) noexcept {
    if (val >= 0) {
        return to_chars(first, last, static_cast<unsigned __int128>(val));
    }

    if (first == last) [[unlikely]] {
        return std::to_chars_result{.ptr = first, .ec = std::errc::value_too_large};
    }

    *first++ = '-';
    return to_chars(first, last, static_cast<unsigned __int128>(-val));
}
```

Let me know if I overlooked something

**Review by liss-h — CHANGES_REQUESTED** [review]
Thanks for the PR, this is the first review. I'll review again after the changes have been made.

File: CMakeLists.txt (hunk 2 above)
Comment (liss-h) [reply CMakeLists.txt]: Getting rid of this completely would also be good, I'll look into that

Comment (liss-h) [reply CMakeLists.txt]: I found a solution: 
The bug is no longer present in clang-20, so we simply require clang>=20 on aarch64 linux. 

If people do not have access to clang 20 on arm, they can use gcc.

File: src/rdf4cpp/util/Int128.hpp (hunk 5 above)
Comment (mcb5637) [reply src/rdf4cpp/util/Int128.hpp]: well, `-std::numeric_limits<__int128>::min()` is a overflow, but we can work around that

File: src/rdf4cpp/util/Int128.hpp (hunk 4 above)
Comment (mcb5637) [reply src/rdf4cpp/util/Int128.hpp]: well, `CheckedInt` and `fun_checked` are designed for different purposes.
`fun_checked` is supposed to have its return value checked immediately
while `CheckedInt` is supposed to be usable by any algorythm not aware of overflow checks

File: conanfile.py [hunk 6]
```diff
@@ -58,6 +59,14 @@ def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
+    def generate(self):
```
Comment (liss-h) [inline conanfile.py]: left over. Some other things in this file as well

File: src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp [hunk 7]
```diff
@@ -87,6 +88,82 @@ F from_chars(std::string_view s) {
...
+        auto n = std::numeric_limits<uint64_t>::digits10;
+        __int128 d = 1;
+        for (int i = 0; i < n; ++i) {
+            d = d * 10;
+        }
+        return d;
+    }();
+    static_assert(max_pow10 == 10'000'000'000'000'000'000ull);
+
+    if (s == "-170141183460469231731687303715884105728") {
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp]: `[[unlikely]]`

Comment (liss-h) [reply src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp]: This check can be bypassed by adding a zero after the minus sign

File: src/rdf4cpp/datatypes/xsd/Decimal.cpp [hunk 8]
```diff
@@ -70,7 +69,7 @@ nonstd::expected<capabilities::Numeric<xsd_decimal>::div_result_cpp_type, Dynami
         return nonstd::make_unexpected(DynamicError::DivideByZero);
     }
 
-    auto r = lhs.div_checked(rhs, 1000);
+    auto r = lhs.div_checked(rhs, 20);
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/Decimal.cpp]: please define a constant somewhere for this. Without context I have no idea why it has to be 20

File: src/rdf4cpp/datatypes/xsd/integers/signed/Integer.cpp [hunk 9]
```diff
@@ -3,32 +3,46 @@
...
 bool capabilities::Logical<xsd_integer>::effective_boolean_value(cpp_type const &value) noexcept {
     return value != 0;
 }
 
+template<>
+nonstd::expected<capabilities::Numeric<xsd_integer>::add_result_cpp_type, DynamicError> capabilities::Numeric<xsd_integer>::add(cpp_type const &lhs, cpp_type const &rhs) noexcept {
+    // https://www.w3.org/TR/xpath-functions/#op.numeric
+    // needs overflow protection
+    cpp_type r;
+    if (rdf4cpp::util::detail::add_checked<rdf4cpp::util::detail::OverflowMode::Checked>(lhs, rhs, r))
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/integers/signed/Integer.cpp]: missing braces (see our coding standard), also in other places in this PR

File: tests/datatype/tests_NumOpResults.cpp [hunk 10]
```diff
@@ -48,10 +50,7 @@ TEST_SUITE("numeric op results") {
 
             // "On underflow, 0.0 must be returned."
             if (min != 0) {  // if min == 0 -> unlimited
-                CHECK(Decimal::sub(min, min) == zero);
-                CHECK(Decimal::div(min, two) == zero);
```
Comment (liss-h) [inline tests/datatype/tests_NumOpResults.cpp:52]: Why did you remove these checks?

File: src/rdf4cpp/BigDecimal.hpp [hunk 11]
```diff
@@ -240,7 +180,7 @@ struct BigDecimal {
         if (significand == 0) {
             return;
         }
-        while (significand & 1) {
+        while ((significand & 1) == 0) {
```
Comment (liss-h) [inline src/rdf4cpp/BigDecimal.hpp]: was this incorrect before?

File: src/rdf4cpp/BigDecimal.hpp [hunk 12]
```diff
@@ -250,22 +190,50 @@ struct BigDecimal {
...
+public:
+    /**
+     * converts a double to a BigDecimal.
+     * this conversion might not be exact, due to the built in limitations of doubles.
+     * if you have the possibility, use one of the other constructors.
+     * @param value
+     * @throw std::overflow_error on exceeding the types numeric limits
+     */
+    explicit BigDecimal(double value) : unscaled_value(0), exponent(0) {
+        if constexpr (!std::numeric_limits<UnscaledValue_t>::is_bounded) {
```
Comment (liss-h) [inline src/rdf4cpp/BigDecimal.hpp]: tbh, I would just remove the template parameters if you need to add special cases for different instantiations

File: src/rdf4cpp/util/Int128.hpp [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+        }
+        template<OverflowMode m, std::size_t MinBits, std::size_t MaxBits, boost::multiprecision::cpp_integer_type SignType, typename Alloc>
+        static constexpr bool add_checked(cpp_int_checked<MinBits, MaxBits, SignType, Alloc> const &a,
+                                          cpp_int_checked<MinBits, MaxBits, SignType, Alloc> const &b,
+                                          cpp_int_checked<MinBits, MaxBits, SignType, Alloc> &result) noexcept {
+            if (m == OverflowMode::Checked) {
+                try {
+                    result = a + b;
+                } catch (std::overflow_error const &) {
+                    return true;
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: these return values are kind of the wrong way around for the function name. That should at least be document (maybe with a comment at the top of the file)

File: src/rdf4cpp/BigDecimal.hpp [hunk 14]
```diff
@@ -727,18 +703,20 @@ struct BigDecimal {
...
-                return std::strong_ordering::less;  // t does fit into the same precision, while o does not
-            if (mul_checked<OverflowMode::Checked>(o, b, o))
-                return std::strong_ordering::less;  // t does fit into the same precision, while o does not
-        } else if (this->exponent < other.exponent) {
-            UnscaledValue_t b{base};
-            if (pow_checked<OverflowMode::Checked>(b, other.exponent - this->exponent, b))
-                return std::strong_ordering::greater;  // o does fit into the same precision, while t does not
-            if (mul_checked<OverflowMode::Checked>(t, b, t))
-                return std::strong_ordering::greater;  // o does fit into the same precision, while t does not
+        if (t != 0 && o != 0) {  // if one of the decimals is 0, compare directly (0*base^e is 0 for all e)
```
Comment (liss-h) [inline src/rdf4cpp/BigDecimal.hpp]: the comment does not match the logic. The condition just says that either must not be null, both could be non-null

File: src/rdf4cpp/util/Int128.hpp [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+                }
+                return false;
+            } else {
+                using Unc = cpp_int_unchecked<MinBits, MaxBits, SignType, Alloc>;
+                result = cpp_int_checked<MinBits, MaxBits, SignType, Alloc>{Unc{a} + Unc{b}};
+                return false;
+            }
+        }
+        template<OverflowMode m, std::size_t MinBits, std::size_t MaxBits, boost::multiprecision::cpp_integer_type SignType, typename Alloc>
+        static constexpr bool add_checked(cpp_int_unchecked<MinBits, MaxBits, SignType, Alloc> const &a,
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: Are these functions with `cpp_int_unchecked` needed? I didn't find any usages with a quick search

Comment (liss-h) [reply src/rdf4cpp/util/Int128.hpp]: A quick test on the branch revealed that they are indeed all unused

File: src/rdf4cpp/util/Int128.hpp [hunk 16]
```diff
@@ -0,0 +1,425 @@
+#ifndef RDF4CPP_INT128_HPP
+#define RDF4CPP_INT128_HPP
+
+#include <boost/multiprecision/cpp_int.hpp>
+#include <rdf4cpp/writer/BufWriter.hpp>
+#include <dice/hash.hpp>
+
+namespace rdf4cpp::util {
+#ifdef __SIZEOF_INT128__
+    using Int128 = __int128;
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp:12]: `Int128` should not be in util namespace.

Just `rdf4cpp::Int128`.
Also for use in `xsd:decimal` define `using Decimal128 = BigDecimal<>` and use that instead

**Review by liss-h — CHANGES_REQUESTED** [review]
Thanks for the changes, I have some more comments

File: tests/datatype/tests_NumOpResults.cpp (hunk 10 above)
Comment (mcb5637) [reply tests/datatype/tests_NumOpResults.cpp:52]: XPath (and by extension SPARQL) do not define underflow themselves. But from how they use it, (see https://www.w3.org/TR/xpath-functions/#op.numeric ) it seems, they understand it the same as it is used in IEEE floats.
And for these (as far as i can tell, i don't seem to have access to the standard) ( https://ieeemilestones.ethw.org/w/images/5/54/Handbook_Floating_Point_Chapter_3.pdf page 31) underflow occurs if the absolute of the result is not 0 but smaller than the smallest representable value.

In which case it makes a lot more sense to return 0.

TLDR: they don't test what they are supposed to test.

File: src/rdf4cpp/BigDecimal.hpp (hunk 11 above)
Comment (mcb5637) [reply src/rdf4cpp/BigDecimal.hpp]: yes, although it only affects some very rare circumstances.

File: src/rdf4cpp/BigDecimal.hpp (hunk 12 above)
Comment (mcb5637) [reply src/rdf4cpp/BigDecimal.hpp]: Its only this conversion. And both are used.

**Comment by liss-h** [comment]
blocked by BETA

**Review by liss-h — CHANGES_REQUESTED** [review]
Please add some additional parsing tests for `Int128` (mostly larger numbers, or large negative numbers) so that we can be sure that it actually works correctly

File: src/rdf4cpp/util/Int128.hpp [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+
+#include <boost/multiprecision/cpp_int.hpp>
+#include <dice/hash.hpp>
+#include <rdf4cpp/writer/BufWriter.hpp>
+
+namespace rdf4cpp {
+#ifdef __SIZEOF_INT128__
+    using Int128 = __int128;
+#else
+    using Int128 = boost::multiprecision::checked_int128_t;
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: I don't recall if we talked about this before but I don't think we need to support platforms where __int128 does not exist.

Comment (liss-h) [reply src/rdf4cpp/util/Int128.hpp]: So we can get rid of all the `#ifdef __SIZEOF_INT128__`

File: src/rdf4cpp/util/Int128.hpp [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+            static constexpr bool mul_checked(T const &a, T const &b, T &result) noexcept {
+                if constexpr (m == OverflowMode::Checked) {
+                    return __builtin_mul_overflow(a, b, &result);
+                } else {
+                    result = a * b;
+                    return false;
+                }
+            }
+            template<OverflowMode m, std::size_t MinBits, std::size_t MaxBits, boost::multiprecision::cpp_integer_type SignType, typename Alloc>
+            static constexpr bool mul_checked(cpp_int_checked<MinBits, MaxBits, SignType, Alloc> const &a,
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: Please move all this boost stuff to a separate header

**Review by liss-h — CHANGES_REQUESTED** [review]
Hi, this review was so long ago that I barely remember anything.
I will probably have to go through everything again, but I have atleast some initial comments for you.

File: src/rdf4cpp/util/Int128.hpp [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+                }
+                result = static_cast<To>(f);
+                return false;
+            }
+        }  // namespace detail
+    }  // namespace util
+}  // namespace rdf4cpp
+
+template<typename Policy>
+struct dice::hash::dice_hash_overload<Policy, __int128> {
```
Comment (liss-h) [inline src/rdf4cpp/util/Int128.hpp]: use https://github.com/dice-group/dice-hash/releases/tag/v0.4.12

**Comment by liss-h** [comment]
POBR breaking change but develop already contains a bump

File: tests/serializer/tests_Serialize.cpp [hunk 20]
```diff
@@ -62,7 +62,7 @@ TEST_SUITE("Serialize") {
             run_ser_test(lit, "\"simple\"");
         }
         SUBCASE("non-inlined xsd:integer") {
-            auto lit = Literal::make_typed_from_value<xsd::Integer>(xsd::Integer::cpp_type{"87960930222089"});
+            auto lit = Literal::make_typed<xsd::Integer>("87960930222089");
```
Comment (nkaralis) [inline tests/serializer/tests_Serialize.cpp]: why is this needed? is the result different?

File: src/rdf4cpp/datatypes/xsd/Decimal.cpp [hunk 21]
```diff
@@ -71,7 +71,7 @@ nonstd::expected<capabilities::Numeric<xsd_decimal>::div_result_cpp_type, Dynami
         return nonstd::make_unexpected(DynamicError::DivideByZero);
     }
 
-    auto r = lhs.div_checked(rhs, 1000);
+    auto r = lhs.div_checked(rhs, cpp_type::default_max_scale_increase);
```
Comment (nkaralis) [inline src/rdf4cpp/datatypes/xsd/Decimal.cpp]: according to the standard this should not use `RoundingMode::Floor` (truncated or rounded). please double check

File: src/rdf4cpp/BigDecimal.hpp [hunk 22]
```diff
@@ -1,931 +1,898 @@
...
+                            throw InvalidNode{"http://www.w3.org/2001/XMLSchema#decimal parsing error: more than one . found"};
+                        }
+                        decimal = true;
+                        continue;
+                    }
+                    if (c < '0' || c > '9') {
+                        throw InvalidNode{"http://www.w3.org/2001/XMLSchema#decimal parsing error: non-numeric char found"};
+                    }
+                    auto n = c - '0';
+                    if (detail::mul_checked<OverflowMode::Checked>(unscaled_value, UnscaledValue_t{base}, unscaled_value)) {
```
Comment (nkaralis) [inline src/rdf4cpp/BigDecimal.hpp]: considering trailing zeros and the canonical .0 here leads to overflows and representable numbers are rejected

File: src/rdf4cpp/BigDecimal.hpp [hunk 23]
```diff
@@ -1,931 +1,898 @@
...
+            /**
+             * division of two BigDecimals.
+             * checks overflow and division by 0.
+             * @param other
+             * @param max_scale_increase after this many decimal places have been added, stop dividing and round
+             * @param mode rounding mode to use
+             * @return
+             */
+            [[nodiscard]] constexpr nonstd::expected<BigDecimal, DecimalError> div_checked(BigDecimal const &other, Exponent_t max_scale_increase, RoundingMode mode = RoundingMode::Floor) const noexcept {
+                if (other.unscaled_value == 0)
```
Comment (nkaralis) [inline src/rdf4cpp/BigDecimal.hpp:615]: please check the following test cases
```
SUBCASE("scaling the dividend must not overflow when the result fits") {
    Dec const max{std::numeric_limits<rdf4cpp::Int128>::max(), 0};
    CHECK(max.div_checked(Dec{"2.5"}, 20) == Dec{"68056473384187692692674921486353642290.8"});
    CHECK(Dec{"170000000000000000000000000000000000000"}.div_checked(Dec{"0.5"}, 20) == Dec{"340000000000000000000000000000000000000"});
    CHECK(Dec{"100000000000000000000000000000000000000"}.div_checked(Dec{"0.25"}, 20) == Dec{"400000000000000000000000000000000000000"});
}
```

**Review by nkaralis — COMMENTED** [review]
I made some comments in BigDecimal.hpp

File: src/rdf4cpp/BigDecimal.hpp [hunk 24]
```diff
@@ -1,931 +1,898 @@
...
+                return unscaled_value >= 0;
+            }
 
-    [[nodiscard]] constexpr UnscaledValue_t get_unscaled_value() const noexcept {
-        return unscaled_value;
-    }
+        private:
+            // op_checked has to be add_checked or sub_checked with the same OverflowMode
+            template<OverflowMode m, bool (*op_checked)(UnscaledValue_t const &t, UnscaledValue_t const &o, UnscaledValue_t &result)>
+            constexpr bool add_or_sub(BigDecimal const &other, BigDecimal &result) const noexcept {
```
Comment (nkaralis) [inline src/rdf4cpp/BigDecimal.hpp:305]: wide intermediate results make the numeric operations overflow, even if the final result would fit? can we use wider types for this cases only?

File: src/rdf4cpp/datatypes/xsd/Decimal.cpp [hunk 25]
```diff
@@ -185,6 +179,54 @@ capabilities::Inlineable<xsd_decimal>::cpp_type capabilities::Inlineable<xsd_dec
     auto const exponent = util::unpack_integral<uint32_t, InlinedDecimal::exponent_size>(data.exponent);
     return cpp_type{unscaled_value, exponent};
 }
+
+template<>
+template<>
+capabilities::Promotable<xsd_decimal>::promoted_cpp_type<0> capabilities::Promotable<xsd_decimal>::promote<0>(cpp_type const &value) noexcept {
```
Comment (nkaralis) [inline src/rdf4cpp/datatypes/xsd/Decimal.cpp:185]: `from_string` throws for small values (`xsd:decimal 1e-60  cast to xsd:float`). this should not be `noexcept`

File: src/rdf4cpp/datatypes/xsd/Decimal.cpp [hunk 26]
```diff
@@ -185,6 +179,54 @@ capabilities::Inlineable<xsd_decimal>::cpp_type capabilities::Inlineable<xsd_dec
...
+    // 19.1.2.1 Casting to xs:float
+    // If ST is xs:decimal then TV is xs:float(SV cast as xs:string) and the conversion is complete.
+    // https://www.w3.org/TR/xpath-functions/#casting-to-float
+    auto const str = writer::StringWriter::oneshot([&value](auto &w) {
```
Comment (nkaralis) [inline src/rdf4cpp/datatypes/xsd/Decimal.cpp]: how long can this string get? do we need to do something against large strings?

**Review by nkaralis — COMMENTED** [review]
Thank you for the latest changes. Two comments from my side

File: src/rdf4cpp/datatypes/xsd/Decimal.cpp (hunk 26 above)
Comment (liss-h) [reply src/rdf4cpp/datatypes/xsd/Decimal.cpp]: decimal is limited so the string is limited.

**Review by nkaralis — APPROVED** [review]
Thank you
